app/models/aaib_report.rb

```diff
@@ -0,0 +1,33 @@
+class AaibReport < Document
+
+  validates :date_of_occurrence, presence: true, date: true, unless: ->(report) {
+    report.report_type == "safety-study" && report.date_of_occurrence.blank?
+  }
+
+  FORMAT_SPECIFIC_FIELDS = [
+    :date_of_occurrence,
+    :aircraft_category,
+    :report_type,
+    :location,
+    :aircraft_type,
+    :registration,
+  ]
+
+  attr_accessor *FORMAT_SPECIFIC_FIELDS
+
+  def initialize(params = {})
+    super(params, FORMAT_SPECIFIC_FIELDS)
```

Comment:
I reckon that `attr_accessor *self.all_fields` would work (assuming it comes after you define `all_fields` in the subclass.  Maybe you'll want:

```ruby
class Document
  def self.common_fields
    [:id, :name, :etc...]
  end
  attr_accessor *self.common_fields

  def self.custom_fields
    []
  end

  def self.all_fields
    common_fields + custom_fields
  end
end

class FooBar < Document
  def self.custom_fields
    [:my, :extra, :fields]
  end
  attr_accessor *self.custom_fields
end
```

Of course, it's annoying to have to remember to override custom_fields *and* provide the attr_accessor so I'd maybe start thinking about some kind of `has_custom_fields` method that meta-programs it all for you.  Getting a bit magical though.  It's a tradeoff between the things you can forget when implementing a new subclass and how complicated the meta-programming is.